xmtp_mls/src/groups/mls_ext/intent.rs

```diff
@@ -0,0 +1,91 @@
+use crate::groups::IntentError;
+use crate::{
+    client::XmtpMlsLocalContext,
+    groups::{
+        intents::{
+            KeyUpdateIntent, SendMessageIntentData, UpdateAdminListIntentData,
+            UpdateGroupMembershipIntent, UpdateMetadataIntentData, UpdatePermissionIntentData,
+        },
+        scoped_client::ScopedGroupClient,
+        GroupError,
+    },
+};
+use openmls::group::MlsGroup;
+use openmls::prelude::Extensions;
+use xmtp_db::{group_intent::IntentKind, XmtpOpenMlsProvider};
+
+pub async fn parse_intent(
+    kind: &IntentKind,
+    data: &[u8],
+    provider: &XmtpOpenMlsProvider,
+    client: impl ScopedGroupClient,
+    group: &MlsGroup,
+) -> Result<Box<dyn GroupIntent>, GroupError> {
+    match kind {
+        IntentKind::UpdateGroupMembership => {
+            let intent =
+                UpdateGroupMembershipIntent::from_stored_bytes(data, provider, client, group)
+                    .await?;
+            Ok(Box::new(intent))
+        }
+        IntentKind::SendMessage => {
+            let intent = SendMessageIntentData::from_bytes(data)?;
+            Ok(Box::new(intent))
+        }
+        IntentKind::KeyUpdate => Ok(Box::new(KeyUpdateIntent)),
+        IntentKind::MetadataUpdate => {
+            let intent = UpdateMetadataIntentData::try_from(data.to_vec())?;
+            Ok(Box::new(intent))
+        }
+        IntentKind::UpdateAdminList => {
+            let intent = UpdateAdminListIntentData::try_from(data.to_vec())?;
+            Ok(Box::new(intent))
+        }
+        IntentKind::UpdatePermission => {
+            let intent = UpdatePermissionIntentData::try_from(data.to_vec())?;
+            Ok(Box::new(intent))
+        }
+    }
+}
+
+#[derive(Debug, derive_builder::Builder)]
```

Comment:
Is the goal to use this `derive_builder` and builder pattern in more places? Seems pretty neat.

Reply:
absolutely! It makes construction of many of our type easier, and lets us share more fields between common constructions